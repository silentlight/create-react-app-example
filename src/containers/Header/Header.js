import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { NavLink } from 'components';

import {
  authenticationSelectors,
  authenticationOperations,
} from 'modules/authentication';

import './Header.css';

class Header extends PureComponent {

  handleClick = () => {
    this.props.logout();
  };

  renderLinks = () => {
    if(this.props.authenticated) {
      return [
        <li key="dashboard"><NavLink to="/dashboard">Dashboard</NavLink></li>,
        <li key="posts"><NavLink to="/posts">Posts</NavLink></li>,
        <li key="logout"><NavLink to="#" onClick={this.handleClick}>Logout</NavLink></li>,
      ];
    }
    else {
      return [
        <li key="home"><NavLink to="/">Home</NavLink></li>,
        <li key="login"><NavLink to="/login">Login</NavLink></li>,
      ]
    }
  };

  render() {
    return (
      <ul className="Header">
        {this.renderLinks()}
      </ul>
    )
  }
}

Header.propTypes = {
  authenticated: PropTypes.bool,
  logout: PropTypes.func,
};

const mapStateToProps = (state, ownProps) => {
  return {
    authenticated: authenticationSelectors.isAuthenticated()(state),
  }
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    logout: () => dispatch(authenticationOperations.logoutRequest()),
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);