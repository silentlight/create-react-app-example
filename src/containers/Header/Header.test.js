import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { Header } from './Header';

describe('Header', () => {

  const defaultProps = {
    authenticated: true,
    logout: jest.fn(),
  };

  let component, props;

  const createComponent = (overrides = {}) => {
    props = Object.assign({}, defaultProps, overrides);

    return shallow(
      <Header {...props } />
    );
  };

  it('renders and matches the snapshot', () => {
    const tree = renderer.create(<Header {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('authenticated prop set to true', () => {

    beforeEach(() => {
      component = createComponent({ authenticated: true });
    });

    it('show authenticated menu when true', () => {
      expect(component.find('li').length).toEqual(3);
    });

  });

  describe('authenticated prop set to false', () => {

    beforeEach(() => {
      component = createComponent({ authenticated: false });
    });

    it('show public menu when false', () => {
      expect(component.html()).toContain('Home');
      expect(component.html()).toContain('Login');
    });

  });

  describe('logout prop', () => {

    beforeEach(() => {
      component = createComponent({ authenticated: true });
    });

    it('logout is called when logout button is clicked', () => {
      component.find('NavLink').last().simulate('click');
      expect(props.logout.mock.calls.length).toEqual(1);
    });

  });

});