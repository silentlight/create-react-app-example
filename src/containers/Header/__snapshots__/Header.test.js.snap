// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders and matches the snapshot 1`] = `
<ul
  className="Header"
>
  <li>
    <a
      className="NavLink"
      onClick={[Function]}
      style={Object {}}
    >
      Dashboard
    </a>
  </li>
  <li>
    <a
      className="NavLink"
      onClick={[Function]}
      style={Object {}}
    >
      Posts
    </a>
  </li>
  <li>
    <a
      className="NavLink"
      onClick={[Function]}
      style={Object {}}
    >
      Logout
    </a>
  </li>
</ul>
`;
